ring/partition_instance_ring.go

```diff
@@ -0,0 +1,123 @@
+package ring
+
+import (
+	"fmt"
+	"time"
+)
+
+type PartitionRingReader interface {
+	PartitionRing() *PartitionRing
+}
+
+// PartitionInstanceRing holds a partitions ring and a instances ring, and provide functions
+// to look up the intersection of the two (e.g. healthy instances by partition).
+type PartitionInstanceRing struct {
+	// partitionsRingReader and partitionsRing are mutually exclusive.
+	partitionsRingReader PartitionRingReader
+	partitionsRing       *PartitionRing
```

Comment:
Sinnce `PartitionRingReader` is an interface, instead of having `partitionsRing` field, we could provide another implementation of `PartitionRingReader` that returns given partition ring. Then there's no need for this mutual-exclusivity.

Reply:
Good idea. Done in https://github.com/grafana/dskit/pull/479/commits/b457572c7c8cbf10b4e7cce98ae3180920821c91